source/request-response/protocol_adapter.c

```diff
@@ -31,66 +34,454 @@
+static void s_aws_mqtt_protocol_adapter_subscription_op_data_destroy(
+    struct aws_mqtt_protocol_adapter_subscription_op_data *subscribe_data) {
+    aws_weak_ref_release(subscribe_data->callback_ref);
+    aws_byte_buf_clean_up(&subscribe_data->topic_filter);
 
-struct aws_mqtt_protocol_adapter_5_impl {
+    aws_mem_release(subscribe_data->allocator, subscribe_data);
+}
+
+struct aws_mqtt_protocol_adapter_publish_op_data {
+    struct aws_allocator *allocator;
+    struct aws_weak_ref *callback_ref;
+
+    void (*completion_callback_fn)(int, void *);
+    void *user_data;
+};
+
+static struct aws_mqtt_protocol_adapter_publish_op_data *s_aws_mqtt_protocol_adapter_publish_op_data_new(
+    struct aws_allocator *allocator,
+    const struct aws_protocol_adapter_publish_options *publish_options,
+    struct aws_weak_ref *callback_ref) {
+    struct aws_mqtt_protocol_adapter_publish_op_data *publish_data =
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_protocol_adapter_publish_op_data));
+
+    publish_data->allocator = allocator;
+    publish_data->callback_ref = aws_weak_ref_acquire(callback_ref);
+    publish_data->completion_callback_fn = publish_options->completion_callback_fn;
+    publish_data->user_data = publish_options->user_data;
+
+    return publish_data;
+}
+
+static void s_aws_mqtt_protocol_adapter_publish_op_data_destroy(
+    struct aws_mqtt_protocol_adapter_publish_op_data *publish_data) {
+    aws_weak_ref_release(publish_data->callback_ref);
+
+    aws_mem_release(publish_data->allocator, publish_data);
+}
+
+/*****************************************************************************************************************/
+
+struct aws_mqtt_protocol_adapter_311_impl {
     struct aws_allocator *allocator;
     struct aws_mqtt_protocol_adapter base;
     struct aws_weak_ref *callback_ref;
     struct aws_mqtt_protocol_adapter_options config;
 
     struct aws_event_loop *loop;
-    struct aws_mqtt5_client *client;
-    struct aws_mqtt5_listener *listener;
+    struct aws_mqtt_client_connection *connection;
+    struct aws_mqtt311_listener *listener;
 };
 
-static void s_aws_mqtt_protocol_adapter_5_destroy(void *impl) {
-    struct aws_mqtt_protocol_adapter_5_impl *adapter = impl;
+static void s_aws_mqtt_protocol_adapter_311_destroy(void *impl) {
+    struct aws_mqtt_protocol_adapter_311_impl *adapter = impl;
 
     // all the real cleanup is done in the listener termination callback
-    aws_mqtt5_listener_release(adapter->listener);
+    aws_mqtt311_listener_release(adapter->listener);
 }
 
-// used by both subscribe and unsubscribe
-struct aws_mqtt_protocol_adapter_5_subscription_op_data {
-    struct aws_allocator *allocator;
+/* Subscribe */
 
-    struct aws_byte_buf topic_filter;
-    struct aws_weak_ref *callback_ref;
+static void s_protocol_adapter_311_subscribe_completion(
+    struct aws_mqtt_client_connection *connection,
```

Comment:
do connection, topic, and packet_id need to be part of this function call? It doesn't appear like they're used in any context and could be removed and added if they're ever needed in the future.

Reply:
Scratch that, we're passing the function to the 311 subscribe callback -_-;;